lib/fluent/plugin/out_datadog.rb

```diff
@@ -56,147 +72,303 @@ def multi_workers_ready?
+    event
+  end
+
+  def max(a, b)
+    a > b ? a : b
+  end
+
+  # Format batch of http events
+  def format_http_event_batch(events)
+    "[#{events.join(',')}]"
+  end
+
+  # Enrich records with metadata such as service, tags or source
+  def enrich_record(tag, time, record)
+    if @dd_sourcecategory
+      record["ddsourcecategory"] ||= @dd_sourcecategory
+    end
+    if @dd_source
+      record["ddsource"] ||= @dd_source
+    end
+    if @dd_tags
+      record["ddtags"] ||= @dd_tags
+    end
+    if @service
+      record["service"] ||= @service
+    end
+    if @dd_hostname
+      # set the record hostname to the configured dd_hostname only
+      # if the record hostname is empty, ensuring having a hostname set
+      # even if the record doesn't contain any.
+      record["hostname"] ||= @dd_hostname
+    end
+
+    if @include_tag_key
+      record[@tag_key] = tag
+    end
+    # If @timestamp_key already exists, we don't overwrite it.
+    if @timestamp_key and record[@timestamp_key].nil? and time
+      record[@timestamp_key] = Time.at(time).utc.iso8601(3)
+    end
+
+    container_tags = get_container_tags(record)
+    unless container_tags.empty?
+      if record["ddtags"].nil? || record["ddtags"].empty?
+        record["ddtags"] = container_tags
+      else
+        record["ddtags"] = record["ddtags"] + "," + container_tags
       end
-      if @service
-        record["service"] ||= @service
+    end
+    record
+  end
+
+  # Compress logs with GZIP
+  def gzip_compress(payload, compression_level)
+    gz = StringIO.new
+    gz.set_encoding("BINARY")
+    z = Zlib::GzipWriter.new(gz, compression_level)
+    begin
+      z.write(payload)
+    ensure
+      z.close
+    end
+    gz.string
+  end
+
+  # Build a new transport client
+  def new_client(logger, api_key, use_http, use_ssl, no_ssl_validation, host, ssl_port, port, use_compression)
+    if use_http
+      DatadogHTTPClient.new logger, use_ssl, no_ssl_validation, host, ssl_port, port, use_compression, api_key
+    else
+      DatadogTCPClient.new logger, use_ssl, no_ssl_validation, host, ssl_port, port
+    end
+  end
+
+  # Top level class for datadog transport clients, managing retries and backoff
+  class DatadogClient
+    def send_retries(payload, max_retries, max_backoff)
+      backoff = 1
+      retries = 0
+      begin
+        send(payload)
+      rescue RetryableError => e
+        if retries < max_retries || max_retries < 0
+          @logger.warn("Retrying ", :exception => e, :backtrace => e.backtrace)
+          sleep backoff
+          backoff = 2 * backoff unless backoff > max_backoff
+          retries += 1
+          retry
+        end
       end
-      if @dd_hostname
-        # set the record hostname to the configured dd_hostname only
-        # if the record hostname is empty, ensuring having a hostname set
-        # even if the record doesn't contain any.
-        record["hostname"] ||= @dd_hostname
+    end
+
+    def send(payload)
+    end
```

Comment:
I guess some exception should raise here to prevent mistakenly use.

Reply:
Good point, updated